Logging Improvements

hummeltech opened the pull request:
* Using [`GLib Logging Framework`](https://developer.gnome.org/programming-guidelines/stable/logging.html.en) for logging
  * Created new `g_logger` function for logging
    * Outputs to `stdout`/`stderr` only when running in `foreground`
      * `stderr` for `message`, `warning`, `critical` & `error` levels
      * `stdout` for `debug` & `info` levels
        * Use `G_MESSAGES_DEBUG=all` environment to enable `debug` to print
    * Otherwise, output will be to `syslog` or `systemd journal` (when appropriate)
* Standardized usage of `{LOG_PRIORITY}: ` prefix in log messages
  * Only when using `syslog`, otherwise `GLib Logging` will take care of it
* Changed `fprintf(stderr`, `printf` & `perror` calls to `g_logger` calls
  * You might want to check them out closely to make sure I chose the right levels
  * No changes to `logging/output` were made to "`foreground`" programs (I.E. `render_*`)
* Changed `0`,`1` to `no_argument`,`required_argument` in `getopt_long`'s `long_options`
  * Fixed `renderd`'s `foreground` opt (should be `no_argument` [0] rather than `reguired_argument` [1])
* Basic test for `mod_tile` module
* ~~Extended `renderd` log priority onto Mapnik's logger~~

xamanu commented:
Personally, I would like to see we introduce logging to the standard error, instead of relying on `syslog` (https://github.com/openstreetmap/mod_tile/issues/219). This PR goes in a good direction, but can we extend it to consolidate logging more?

hummeltech commented:
Yeah, you make a good point in your issue, I think I can work on something like that.

xamanu commented:
Great, @hummeltech, many thanks. Do you want to do this in this PR or in another one?

hummeltech commented:
I'll update this pull request I think.

xamanu commented:
Today, I merged the changes for code formatting (#228), this will require a rebase on the latest `master`. This will cause conflicts because the style has changed, but these should be easy to fix, as no logic has changed. Improving the logging is key for `mod_tile`. Thanks for contributing to that!

hummeltech commented:
Awesome, I wanted to make some formatting changes myself, but didn't include anything to keep this pull request simple. Sounds good.

hummeltech commented:
Alright, I have finishing rebasing this branch, I will continue adding the `stderr` "logging" output in the next day or two!

hummeltech commented:
Sorry, I am a little busy with other work at the moment but based on your comments in the referenced issue, wouldn't it be best to run `renderd` with `--foreground` when being started by systemd? If you were envisioning having another config option to choose the logging backend (I.E. syslog, stdout, stderr, etc) I think that would make the most sense, especially for supporting other Unix-like operating systems. What exactly did you have in mind?

hummeltech commented:
I am working on integrating a library, `zlog`, to handle this and the results so far are promising.

xamanu commented:
Thank you for your work on this!

Yes a systemd service starts `renderd --foreground` but currently still writes logs to syslog.

I can understand that you want to use zlog. However I would prefer not to introduce another dependency that would be needed to be shipped within the program. Please consider [log4c](http://log4c.sourceforge.net) or [glib](https://developer.gnome.org/glib/2.66/glib-Message-Logging.html) instead, that are available in most OS this usually runs on. If you don't want to rely on standard C (print/fprintf) and allow richer options, I would suggest to go with one of these.

hummeltech commented:
Hallo @xamanu, I have pushed up some of the work I have done to unify the logging, take a look and let me know if you are satisfied with the changes. I am not especially experienced with working with C development, so feel free to make any recommendations.

hummeltech commented:
Here's some sample output from my test Docker container:
```
** Message: 01:11:50.463: INFO: Rendering daemon started
** Message: 01:11:50.463: INFO: Initialising request_queue
** Message: 01:11:50.464: INFO: Parsing section renderd
** Message: 01:11:50.464: INFO: Parsing render section 0
** Message: 01:11:50.464: INFO: Parsing section mapnik
** Message: 01:11:50.464: INFO: Parsing section default
** Message: 01:11:50.464: INFO: config renderd: ip socket=tile-backend:7654
** Message: 01:11:50.464: INFO: config renderd: num_threads=4
** Message: 01:11:50.464: INFO: config renderd: num_slaves=0
** Message: 01:11:50.464: INFO: config renderd: tile_dir=/var/lib/mod_tile
** Message: 01:11:50.464: INFO: config renderd: stats_file=/var/run/renderd/renderd.stats
** Message: 01:11:50.464: INFO: config mapnik:  plugins_dir=/usr/lib64/mapnik/input
** Message: 01:11:50.464: INFO: config mapnik:  font_dir=/usr/share/fonts
** Message: 01:11:50.464: INFO: config mapnik:  font_dir_recurse=1
** Message: 01:11:50.464: INFO: config renderd(0): Active
** Message: 01:11:50.464: INFO: config renderd(0): ip socket=tile-backend:7654
** Message: 01:11:50.464: INFO: config renderd(0): num_threads=4
** Message: 01:11:50.464: INFO: config renderd(0): tile_dir=/var/lib/mod_tile
** Message: 01:11:50.464: INFO: config renderd(0): stats_file=/var/run/renderd/renderd.stats
** Message: 01:11:50.464: INFO: config map 0:   name(default) file(/etc/openstreetmap-carto/tile-backend.xml) uri(/osm_tiles/) htcp() host(127.0.0.1)
** Message: 01:11:50.464: INFO: Initialising TCP/IP server socket on tile-backend:7654
** Message: 01:11:50.464: INFO: Renderd is using mapnik version 3.0.23
```

You can enable debug messages with the `G_MESSAGES_DEBUG` environment variable, I.E. `G_MESSAGES_DEBUG=all`:
```
** (process:8): DEBUG: 01:11:51.398: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSinhalaUI-Thin.ttf
** (process:8): DEBUG: 01:11:51.399: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-Black.ttf
** (process:8): DEBUG: 01:11:51.399: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-Bold.ttf
** (process:8): DEBUG: 01:11:51.399: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-ExtraBold.ttf
** (process:8): DEBUG: 01:11:51.400: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-ExtraLight.ttf
** (process:8): DEBUG: 01:11:51.401: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-Light.ttf
** (process:8): DEBUG: 01:11:51.402: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-Medium.ttf
** (process:8): DEBUG: 01:11:51.403: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-Regular.ttf
** (process:8): DEBUG: 01:11:51.403: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-SemiBold.ttf
** (process:8): DEBUG: 01:11:51.404: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols-Thin.ttf
** (process:8): DEBUG: 01:11:51.405: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSansSymbols2-Regular.ttf
** (process:8): DEBUG: 01:11:51.405: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-Black.ttf
** (process:8): DEBUG: 01:11:51.406: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-Bold.ttf
** (process:8): DEBUG: 01:11:51.406: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-ExtraBold.ttf
** (process:8): DEBUG: 01:11:51.407: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-ExtraLight.ttf
** (process:8): DEBUG: 01:11:51.408: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-Light.ttf
** (process:8): DEBUG: 01:11:51.409: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-Medium.ttf
** (process:8): DEBUG: 01:11:51.410: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-Regular.ttf
** (process:8): DEBUG: 01:11:51.410: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-SemiBold.ttf
** (process:8): DEBUG: 01:11:51.410: DEBUG: Loading font: /usr/share/fonts/google-noto-fonts-extra/NotoSerifTibetan-Thin.ttf
** (process:8): DEBUG: 01:11:51.411: DEBUG: Loading font: /usr/share/fonts/hanazono/HanaMinA.ttf
** (process:8): DEBUG: 01:11:51.412: DEBUG: Loading font: /usr/share/fonts/hanazono/HanaMinB.ttf
** Message: 01:11:51.412: INFO: Running in foreground mode...
** (process:8): DEBUG: 01:11:51.412: DEBUG: Starting stats thread
** (process:8): DEBUG: 01:11:51.412: DEBUG: init_storage_backend: initialising file storage backend at: /var/lib/mod_tile
** (process:8): DEBUG: 01:11:51.412: DEBUG: init_storage_backend: initialising file storage backend at: /var/lib/mod_tile
** (process:8): DEBUG: 01:11:51.412: DEBUG: init_storage_backend: initialising file storage backend at: /var/lib/mod_tile
** (process:8): DEBUG: 01:11:51.412: DEBUG: init_storage_backend: initialising file storage backend at: /var/lib/mod_tile
```

xamanu commented:
Thanks a lot, @hummeltech! Looks great at a first glance. I will give this and the other PR which is open for review a round of testing and a deeper review within the next few days.

xamanu commented:
I'd love to bring this is. It currently has some conflicts, it would be nice to resolve then and test it for the next 0.6 release (#227).

hummeltech commented:
Alright @xamanu, I will resolve these new conflicts.

hummeltech commented:
:100: I think it should be ready to go.

xamanu commented:
Thanks a lot! Unfortunately the git history seems to be messed up and github tells me "This branch cannot be rebased due to conflicts". It would be good to make a clean commit on top of recent `master`. If you can do this, it would be great, if not, I'm happy to reroll this within the next days or few weeks. Once this is done, I would include it into the release candidate build so we can test it comfortably.

hummeltech commented:
Hallo @xamanu, that's strange, I don't know what happened, I have rebased and squashed everything into a single commit, hopefully that helps. Let me know.

Thank you

hummeltech commented on src/store.c:
```diff
@@ -22,44 +22,9 @@
 #include "store_null.h"
+#include "g_logger.h"
 
-//TODO: Make this function handle different logging backends, depending on if on compiles it from apache or something else
```
I think I may have initially misunderstood this TODO comment, the `g_logger` function implementation does pretty much the same as this `log_message` function did. It  does not, however, implement what the TODO describes, I.E. calling `ap_log_*` functions when built into `mod_tile.so`.

xamanu commented:
Submitted this to `develop` and built new debian packages as a version 0.6 release candidate for testing purposes. See instructions: https://deb.openbuildingmap.org/

Please let me know if the work or not. Many thanks.

xamanu commented:
Thanks a lot @hummeltech! I just tested successfully and put it into `master`, after rebasing it again. Closing here.
